spec.bs

```diff
@@ -8159,43 +8321,154 @@ To <dfn>batch or fetch trusted bidding signals</dfn> given a [=trusted bidding s
+    1. Let |compressionGroupMap|[|compressionGroupId|] be an empty [=map=], whose [=map/keys=] are integers
+      and [=map/values=] are [=maps=].
+  1. Set |executionMode| to |group|'s [=interest group/execution mode=].
+  1. If |executionMode| equal to "`group-by-origin`", set |partitionId| to 0.
+  1. Otherwise:
+    1. If |compressionGroupMap|[|compressionGroupId|] [=map/contains=] 0, set |partitionId| to
+      [=map/size=] of |compressionGroupMap|[|compressionGroupId|].
+    1. Otherwise, set |partitionId| to sum [=list/size=] of |compressionGroupMap|[|joiningOrigin|]
+      and 1.
+  1. Set |interestGroupIdMap|[|group|'s [=interest group/name=]] to [=tuple=] of |compressionGroupId|
+    and |partitionId|.
+  1. If |compressionGroupMap|[|compressionGroupId|] does not [=map/contain=] |partitionId|:
+    1. Let |partition| be an empty [=partition=].
+    1. Set |partition|'s [=partition/id=] to |partitionId|.
+    1. Let |namespace| be an empty [=map=], whose [=map/keys=] are [=strings=] and [=map/values=] are [=list=]
+      of [=strings=].
+    1. Set |namespace|["interest_group_names"] to [|group|'s [=interest group/name=]].
+    1. Set |namespace|["bidding_keys"] to |group|'s [=interest group/trusted bidding signals keys=].
+    1. Set |partition|'s [=partition/namespace=] to |namespace|.
+    1. Let |metadata| be an empty [=map=], whose [=map/keys=] and [=map/values=] are [=strings=].
+    1. Set |metadata|["experiment_group_id"] to |experimentGroupId|.
+    1. Set |metadata|[|slotSizeParams|[0]] to |slotSizeParams|[1].
+    1. Set |partition|'s [=partition/metadata=] to |metadata|.
+    1. Set |compressionGroupMap|[|compressionGroupId|][|partitionId|] to |partition|.
+  1. Otherwise:
+    1. [=list/Append=] |group|'s [=interest group/name=] into |compressionGroupMap|[|compressionGroupId|]
+      [|partitionId|]["interest_group_names"].
+    1. [=list/Append=] |group|'s [=interest group/trusted bidding signals keys=] into
+      |compressionGroupMap|[|compressionGroupId|][|partitionId|]["bidding_keys"].
+1. [=map/For each=] |id| → |group| in |compressionGroupMap|:
+  1. Let |compressionGroup| be an empty [=compression group=].
+  1. Set |compressionGroup|'s [=compression group/compression group id=] to |id|.
+  1. Set |compressionGroup|'s [=compression group/partition=] to an empty [=list=].
+  1. [=list/For each=] |partition| in |group|'s [=map/values=]:
+    1. [=list/Append=] |partition| to |compressionGroup|'s [=compression group/partition=].
+  1. [=list/Append=] |compressionGroup| to |compressionGroups|.
+1. Let |metadata| be an empty [=map=], whose [=map/keys=] and [=map/values=] are [=strings=].
+1. Let |hostname| be the result of [=string/UTF-8 percent-encoding=] the
+  [=serialization of an origin|serialized=] |topLevelOrigin| using [=component percent-encode set=].
+1. Set |metadata|["hostname"] to |hostname|.
+1. Let |keyInfo| be the result of [=looking up the server encryption key=] with |owner| and

```

Comment:
the look up algorithm can return failure. IIUC, we'll not generate the request in that case?

Reply:
Yes, from the code's perspective, if it fails to get the public key, there will be a worklet failure.

Reply:
Returning all nulls is fine, or you can return a failure instead, which might be more readable (but then you can no longer do "let (...,...,...) be the result of ...", but need to do "let |..Result| be the result of ...; if |...Result| is failure: ...; Otherwise, let (...,...,...) be |...Result|".

But your `If |body| is null, set |signals| to failure and return.` is not correct. That algorithm returns a list, but this step would return nothing. You can do the same of return failure, or do return (null, ...., ...) with reasonable values for the other two fields. I'd prefer failure if it's a failure. Is there any difference between signals being null VS there's a failure (i.e., can signals be null if there's no failure)?

And what's the term that the public can differentiate v1 and v2? In other words, the algorithm says "trusted key value signals", would people know it's talking about v2? Besides terms v1/v2, is there another term/concept that people can obviously know whether it's v1/v2, like are we only calling v2 "trusted key value signals"?

Reply:
Changed all "set |signals| to failure and return." to "return « null, null, null ».", since the caller need to check if the first return object is null or not. E.g. 
```
1. Let « |partialTrustedBiddingSignals|, |partialPerInterestGroupData|, |dataVersion| » be the
      result of [=fetching trusted key value signals=] with |signalsUrl|, |requestBody|, |context|,
      |scriptOrigin|, |policyContainer|, |interestGroupIdMap| and true.
    1. If |partialTrustedBiddingSignals| is not null:
```

For the naming, I don't think V2 is a official word which we want to put it in the spec, and trusted key value seems to be a good one for V2, we can check with Paul about it.

Reply:
yeah I didn't mean we should use V2, just wanted to ask whether there's a good term that we use to differentiate between v1 and v2. Currently I see you used "trusted key value signals", which I wanted to confirm if that's clear about that the algorithm is about v2 stuff.